crates/nu-protocol/src/engine/engine_state.rs

```diff
@@ -64,23 +64,29 @@ impl Clone for IsDebugging {
 /// will refer to the corresponding IDs rather than their definitions directly. At runtime, this means
 /// less copying and smaller structures.
 ///
+/// Many of the larger objects in this structure are stored within `Arc` to decrease the cost of
+/// cloning `EngineState`. While `Arc`s are generally immutable, they can be modified using
+/// `Arc::make_mut`, which automatically clones to a new allocation if there are other copies of
+/// the `Arc` already in use, but will let us modify the `Arc` directly if we have the only
+/// reference to it.
+///
 /// Note that the runtime stack is not part of this global state. Runtime stacks are handled differently,
 /// but they also rely on using IDs rather than full definitions.
 #[derive(Clone)]
 pub struct EngineState {
-    files: Vec<(String, usize, usize)>,
-    file_contents: Vec<(Vec<u8>, usize, usize)>,
+    files: Vec<(Arc<String>, usize, usize)>,
+    file_contents: Vec<(Arc<Vec<u8>>, usize, usize)>,
     pub(super) virtual_paths: Vec<(String, VirtualPath)>,
     vars: Vec<Variable>,
-    decls: Vec<Box<dyn Command + 'static>>,
-    pub(super) blocks: Vec<Block>,
-    pub(super) modules: Vec<Module>,
+    decls: Arc<Vec<Box<dyn Command + 'static>>>,
+    pub(super) blocks: Vec<Arc<Block>>,
+    pub(super) modules: Vec<Arc<Module>>,
```

Comment:
The inner types are already large to probably benefit from an `Arc` themselves (see the cloning of `Block`s).

Future direction may be to `Arc` the `Vec`. There will be quite a number of elements.

(looking at it further, `Module` itself doesn't get cloned elsewhere)

Reply:
Yeah, agree - I realized this later too, the list could get quite large and we don't want to refcount++ a thousand Arcs every time. I think I'll patch this